docs/design/2022-11-25-global-resource-control.md

```diff
@@ -0,0 +1,240 @@
+
+Moreover, this resource control mechanism can be used by Resource Manage in the future of TiDB Cloud. Users can pay for the actual usage.  So, consider scenarios. It supports two models.
+
+- **Provisioned:** On-Premise deployment, multi-applications shared on the big cluster. In this mode, you provision the number of RU rate limits or(CPU, IO Rate) and priority for the users/tenant/sessions.
+- **Serverless:** In this mode, the user can pay the total cumulative RU for the tenant or user. Then the resources can not be limited until the RU is exhausted. This requirement TiDB Cloud has sufficient flexibility. such as auto-scale resources.
+
+## Detailed Design
+
+![resource control layout.png](imgs/resource-control-layout.png)
+
+**Resource Group** helps users to control over its resources, to enable or restrict resource consumption. resource groups have attributes that define the group. all attributes can be set at group creation time and  modified any time thereafter. The attributes define how many resources unit in this group relative CPU, IO, Memory and so on.
+resource groups can be assigned to different resource objects. such as session/users/ tenant.
+There are two dimensions to control the resources, one is Global Quota dimension, which is more of a tenant-level application, users can plan the use of resources in provision, to limit the resources usage. On the other side is the scheduling(weight ) dimension, to ensure better control behavior when the resources are sufficient.
+**Global Quota dimension:**  This part is the upper limiter and lower limit(Reservation)  for the resource group. This defines the provision of the resource usage of the application layer.  Here mainly use a token bucket algorithm to limit the resources at the entry of TiDB, count total resource usage including sql, storage layer.
+**Scheduling(weight) dimension:**  Currently, it’s mainly used in the internal mechanism, such as inner jobs DDL, analyzed to reduce the background job influence for priority purpose, let background jobs not influence front jobs. and on the other hand, it makes sure the resource group is not abused, prevent a resource group makes one or more nodes overhead.  it may have a user interface in the future, not exposed now for simplicity and clarity . Here mainly use the mclock/dmclock algorithm at the storage layer, and use different priority queues to influence different workers of the SQL layer.
+Will introduce this two dimensions control mechanism in the following.
+
+### User SQL interface
+The `Request Unit` can be assigned to a different level, like user or tenant. The syntax is like MySQL resource group, but it is used differently:
+
+```sql
+CREATE RESOURCE GROUP name resouce_unit=value [, ... ]
+/*eg:*/
+CREATE RESOURCE GROUP small_application RRU=20000, WRU=10000;
```

Comment:
No `RRU/WRU`, replaced by `RU` now?

Reply:
I keep the original copy plus the RRU as well in the RFC. It's not completely the same as the implementation. Is it ok?